internal/services/notification/commitstatusdelivery.go

```diff
@@ -0,0 +1,225 @@
+		err := n.d.Do(ctx, func(tx *sql.Tx) error {
+			var err error
+			commitStatusDeliveries, err = n.d.GetCommitStatusDataDeliveriesAfterSequence(tx, curCommitStatusDeliverySequence, types.DeliveryStatusNotDelivered, MaxCommitStatusDeliveriesQueryLimit)
+			if err != nil {
+				return errors.WithStack(err)
+			}
+
+			return nil
+		})
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		for _, c := range commitStatusDeliveries {
+			if err := n.handleCommitStatusDataDelivery(ctx, c.ID); err != nil {
+				n.log.Err(err).Msgf("failed to trigger commit status data delivery event")
+			}
+
+			curCommitStatusDeliverySequence = c.Sequence
+		}
+
+		if len(commitStatusDeliveries) < MaxCommitStatusDeliveriesQueryLimit {
+			return nil
+		}
+	}
+}
+
+func (n *NotificationService) handleCommitStatusDataDelivery(ctx context.Context, commitStatusDataDeliveryID string) error {
+	var commitStatusData *types.CommitStatusData
+
+	err := n.d.Do(ctx, func(tx *sql.Tx) error {
+		var err error
+
+		commitStatusDataDelivery, err := n.d.GetCommitStatusDataDeliveryByID(tx, commitStatusDataDeliveryID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		if commitStatusDataDelivery.DeliveryStatus != types.DeliveryStatusNotDelivered {
+			return nil
+		}
+
+		commitStatusData, err = n.d.GetCommitStatusDataByID(tx, commitStatusDataDelivery.CommitStatusDataID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	if commitStatusData == nil {
+		return nil
+	}
+
+	commitStatusDelivered := false
+	resp, err := n.updateCommitStatus(ctx, commitStatusData)
+	// err != nil is not checked because every error is considered a failed delivery
+	if err == nil && resp != nil && resp.StatusCode == http.StatusCreated {
+		commitStatusDelivered = true
+	}
+
+	var deliveredAt *time.Time
+	var statusCode int
+
+	if resp != nil {
+		deliveredAt = util.TimeP(time.Now())
+		statusCode = resp.StatusCode
+	}
+
+	err = n.d.Do(ctx, func(tx *sql.Tx) error {
+		var err error
+
+		commitStatusDataDelivery, err := n.d.GetCommitStatusDataDeliveryByID(tx, commitStatusDataDeliveryID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		if commitStatusDataDelivery.DeliveryStatus != types.DeliveryStatusNotDelivered {
+			return nil
+		}
+
+		if commitStatusDelivered {
+			commitStatusDataDelivery.DeliveryStatus = types.DeliveryStatusDelivered
+		} else {
+			commitStatusDataDelivery.DeliveryStatus = types.DeliveryStatusDeliveryError
+		}
+
+		commitStatusDataDelivery.DeliveredAt = deliveredAt
+		commitStatusDataDelivery.StatusCode = statusCode
+
+		if err = n.d.UpdateCommitStatusDataDelivery(tx, commitStatusDataDelivery); err != nil {
+			return errors.WithStack(err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+func (n *NotificationService) updateCommitStatus(ctx context.Context, commitStatusData *types.CommitStatusData) (*http.Response, error) {
```

Comment:
So you're changing the whole gitsources apis to return the http.Response? Then you look at the http.Response and check the status code. This assumes that ALL the gitsources use the same http status code etc... Doing the check here is wrong.

You should check if the commit status has been successfully delivered in the gitsource and then return a common type that contains all the required information (or just a bool if you only need to know if it has been delivered).